File: app/src/main/java/com/tronprotocol/app/plugins/NotesPlugin.java
package com.tronprotocol.app.plugins;

import android.content.Context;

import java.util.ArrayList;
import java.util.List;

/**
 * Simple notes plugin to demonstrate additional plugin extensibility.
 */
public class NotesPlugin implements Plugin {
    private static final String ID = "notes";
    private final List<String> notes = new ArrayList<>();
    private boolean enabled = true;

    @Override
    public String getId() {
        return ID;
    }

    @Override
    public String getName() {
        return "Notes";
    }

    @Override
    public String getDescription() {
        return "Create quick notes using add|text, list, and clear commands";
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    @Override
    public PluginResult execute(String input) {
        long start = System.currentTimeMillis();
        String[] parts = input.split("\\|", 2);
        String command = parts[0].trim().toLowerCase();

        switch (command) {
            case "add":
                if (parts.length < 2 || parts[1].trim().isEmpty()) {
                    return PluginResult.error("Use add|<note>", System.currentTimeMillis() - start);
                }
                notes.add(parts[1].trim());
                return PluginResult.success("Added note. Total notes: " + notes.size(), System.currentTimeMillis() - start);
            case "list":
                if (notes.isEmpty()) {
                    return PluginResult.success("No notes yet", System.currentTimeMillis() - start);
                }
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < notes.size(); i++) {
                    sb.append(i + 1).append(". ").append(notes.get(i)).append("\n");
                }
                return PluginResult.success(sb.toString(), System.currentTimeMillis() - start);
            case "clear":
                notes.clear();
                return PluginResult.success("Cleared all notes", System.currentTimeMillis() - start);
            default:
                return PluginResult.error("Unknown command. Use add|text, list, clear", System.currentTimeMillis() - start);
        }
    }

    @Override
    public void initialize(Context context) {
        // No-op
    }

    @Override
    public void destroy() {
        notes.clear();
    }
}

File: app/src/main/java/com/tronprotocol/app/plugins/NotesPlugin.kt
package com.tronprotocol.app.plugins

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import org.json.JSONArray

/**
 * Notes plugin with persistent storage.
 *
 * Supports creating, listing, searching, deleting, and clearing notes.
 * Notes are persisted across app restarts via SharedPreferences.
 *
 * Commands:
 *   add|<note text>       - Add a new note
 *   list                  - List all notes
 *   search|<query>        - Search notes containing query text
 *   delete|<number>       - Delete note by number (1-based)
 *   clear                 - Delete all notes
 *   count                 - Get the number of notes
 */
class NotesPlugin : Plugin {

    companion object {
        private const val ID = "notes"
        private const val TAG = "NotesPlugin"
        private const val PREFS_NAME = "tronprotocol_notes"
        private const val KEY_NOTES = "notes_list"
    }

    private val notes = mutableListOf<String>()
    private var prefs: SharedPreferences? = null

    override val id: String = ID

    override val name: String = "Notes"

    override val description: String =
        "Manage notes: add|text, list, search|query, delete|number, clear, count. " +
            "Notes are persisted across app restarts."

    override var isEnabled: Boolean = true

    override fun execute(input: String): PluginResult {
        val start = System.currentTimeMillis()
        val parts = input.split("\\|".toRegex(), 2)
        val command = parts[0].trim().lowercase()

        return when (command) {
            "add" -> {
                if (parts.size < 2 || parts[1].trim().isEmpty()) {
                    PluginResult.error("Use add|<note>", System.currentTimeMillis() - start)
                } else {
                    notes.add(parts[1].trim())
                    saveNotes()
                    PluginResult.success(
                        "Added note #${notes.size}: ${parts[1].trim()}",
                        System.currentTimeMillis() - start
                    )
                }
            }
            "list" -> {
                if (notes.isEmpty()) {
                    PluginResult.success("No notes yet.", System.currentTimeMillis() - start)
                } else {
                    val sb = buildString {
                        append("Notes (${notes.size}):\n")
                        notes.forEachIndexed { index, note ->
                            append("${index + 1}. $note\n")
                        }
                    }
                    PluginResult.success(sb.trimEnd(), System.currentTimeMillis() - start)
                }
            }
            "search" -> {
                if (parts.size < 2 || parts[1].trim().isEmpty()) {
                    PluginResult.error("Use search|<query>", System.currentTimeMillis() - start)
                } else {
                    val query = parts[1].trim().lowercase()
                    val matches = notes.mapIndexedNotNull { index, note ->
                        if (note.lowercase().contains(query)) {
                            "${index + 1}. $note"
                        } else null
                    }
                    if (matches.isEmpty()) {
                        PluginResult.success(
                            "No notes matching '$query'.",
                            System.currentTimeMillis() - start
                        )
                    } else {
                        val sb = buildString {
                            append("Found ${matches.size} matching note(s):\n")
                            matches.forEach { append("$it\n") }
                        }
                        PluginResult.success(sb.trimEnd(), System.currentTimeMillis() - start)
                    }
                }
            }
            "delete" -> {
                if (parts.size < 2 || parts[1].trim().isEmpty()) {
                    PluginResult.error("Use delete|<number>", System.currentTimeMillis() - start)
                } else {
                    val index = parts[1].trim().toIntOrNull()
                    if (index == null || index < 1 || index > notes.size) {
                        PluginResult.error(
                            "Invalid note number. Use 1-${notes.size}.",
                            System.currentTimeMillis() - start
                        )
                    } else {
                        val removed = notes.removeAt(index - 1)
                        saveNotes()
                        PluginResult.success(
                            "Deleted note #$index: $removed",
                            System.currentTimeMillis() - start
                        )
                    }
                }
            }
            "clear" -> {
                val count = notes.size
                notes.clear()
                saveNotes()
                PluginResult.success(
                    "Cleared $count note(s).",
                    System.currentTimeMillis() - start
                )
            }
            "count" -> {
                PluginResult.success(
                    "${notes.size} note(s).",
                    System.currentTimeMillis() - start
                )
            }
            else -> PluginResult.error(
                "Unknown command '$command'. Use: add|text, list, search|query, delete|number, clear, count",
                System.currentTimeMillis() - start
            )
        }
    }

    override fun initialize(context: Context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        loadNotes()
        Log.d(TAG, "NotesPlugin initialized with ${notes.size} persisted notes")
    }

    override fun destroy() {
        notes.clear()
        prefs = null
    }

    private fun saveNotes() {
        val json = JSONArray()
        notes.forEach { json.put(it) }
        prefs?.edit()?.putString(KEY_NOTES, json.toString())?.apply()
    }

    private fun loadNotes() {
        val data = prefs?.getString(KEY_NOTES, null) ?: return
        try {
            val json = JSONArray(data)
            notes.clear()
            for (i in 0 until json.length()) {
                notes.add(json.getString(i))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error loading notes", e)
        }
    }
}
